### ISSUE TITLE
Changing optimizer of restored network messes up training output

### ISSUE THREAD

**Issue by cnsmth** [tensorflow/tensorflow#21459]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
 Ubuntu 16.04.2 LTS (GNU/Linux 4.4.0-130-generic x86_64)

- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
N/A

- **TensorFlow installed from (source or binary)**:
binary

- **TensorFlow version (use command below)**:
v1.4.0-rc1-11-g130a514 1.4.0

- **Python version**:
Python 3.5.2

- **Bazel version (if compiling from source)**:
N/A

- **GCC/Compiler version (if compiling from source)**:
N/A

- **CUDA/cuDNN version**:
Cuda compilation tools, release 8.0, V8.0.61

- **GPU model and memory**:
GeForce GTX 1080 Ti (11GB)

- **Exact command to reproduce**:

### Describe the problem
I am trying to train a model to do face detection, using Resnet_v1_50 as a backbone net and restoring it from the checkpoint provided in tf-slim. I am following the paper "MB-FCN for Face Detection" in building my network. Initially trained with gradient descent, the model was converging slowly. When the optimizer was changed from Gradient Descent (which ResNet was trained with) to a different optimizer (momentum, adam) there were wildly different outputs, that were nowhere near correct even though the training and validation losses were lower and care was taken to initialize all weights for new variables (from momentum for example). There seems to perhaps be an issue with the tf Saver class (restoring variables goes wrong) or an issue changing the optimizer of a restored network from the one it was originally trained with. I also had this issue when using a different model for text detection (changing optimizer messing with output). Restoring the weights from the fine-tuned model with another optimizer goes wrong.

### Source code / logs
Code while training from ResNet:
    global_step = tf.train.get_or_create_global_step()
    with slim.arg_scope(resnet_v1.resnet_arg_scope()):
        vars_to_restore = slim.get_variables_to_restore(['resnet_v1_50'])

    optimizer = tf.train.GradientDescentOptimizer(0.001)
    optimizer = optimizer.minimize(loss, global_step=global_step) 

    config = tf.ConfigProto(allow_soft_placement=True)
    config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1
    overall_saver = tf.train.Saver(slim.get_variables_to_restore())

    with tf.Session(config=config) as sess:
        resnet_model_path = 'backbone_net/resnet_v1_50.ckpt'
        vars_to_initialize = []
        all_vars = slim.get_variables_to_restore()
        restored_vars = set(vars_to_restore)
        for var in all_vars:
            if var not in restored_vars:
                vars_to_initialize.append(var)

        restorer = tf.train.Saver(vars_to_restore)
        restorer.restore(sess, resnet_model_path)
        sess.run(tf.variables_initializer(vars_to_initialize))

Line to save weights while training:
        overall_saver.save(sess, 'weights/model.ckpt', global_step=global_step)

Code to restore weights once fine-tuned:
    iterator, init_op = create_dataset('test_imgs.csv', is_training=False)
    b1_cls, b2_cls, b1_reg, b2_reg, ratio_h, ratio_w = mbfcn_model(iterator, is_training=False)

    config = tf.ConfigProto(allow_soft_placement=True)
    config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1

    with tf.Session(config=config) as sess:
        chkpt = 'weights/'
        restorer = tf.train.Saver(slim.get_variables_to_restore())
        restorer.restore(sess, tf.train.latest_checkpoint(chkpt))

**Comment by jackd** [comment]
What do you results look like if you infer using `is_training=True`? Obviously that's not what you -want- to do, but I've found I have similar issues with transfering `slim` model weights to different applications that are resolved by inferring in training mode, which would indicate the problem is likely related to `slim`'s batch normalization - either not saving moving averages, or not loading them correctly.

I haven't got to the bottom of the issue yet, but I can't make a minimal working example which exhibits similar behaviour either... but it sounds like I might not be the only one with such issues.

**Comment by jackd** [comment]
After extensive testing I found my issue was related to the very high value of `decay` used in batch normalization. Lowering the decay term to `0.9` resulted in almost immediate consistency between training performance and evaluation performance. The best explanation I have is that update operations were being run, but the decay value so close to 1 (0.997 for my case using mobilenet) meant moving averages weren't keeping up with the changes to the batch statistics as a result of faster training.

I find this very surprising, but it's the best I've got...

**Comment by jackd** [comment]
One more update: I've found lowering the learning rate/switching to a simpler optimizer (momentum, from Adam) resolved my issues. Again I find this incredibly surprising, but it lends support to the idea that these networks might not have convergent activations, even if the final inferences are relatively stable... Alternatively I'm completely wrong and there's a bug somewhere. /shrug

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
